### PR TITLE
Divide BattleAI on the smaller files.

### PR DESCRIPTION
**Pull request by FeniksFire**
Divide BattleAI on the smaller files.

### CODE REVIEW

**Comment by FeniksFire** [comment]
1. Uhh, I see in my IDE all Tabs, ehh, but you have right on github them somehow doesn't exist.
2. I will try fix this.
3. I don't see before in this files any copyright : https://github.com/vcmi/vcmi/tree/develop/AI/BattleAI, so what I could write? I can describe only yourself, but what about other?
4. Regarding the leaving LOGFL and LOGI you're right, I don't want delete it I thought that I removed only the point of use it because they are wrong/nonsense used, this must be fix it in the future.

**Comment by ArseniyShestakov** [comment]
3. Just look in other part of engine. We have standard copyright header without a year or names.
You can add yourself to AUTHORS as well since we don't include copyright in source files.

File: AI/BattleAI/CMakeLists.txt [hunk 1]
```diff
@@ -5,8 +5,15 @@ include_directories(${Boost_INCLUDE_DIRS} ${CMAKE_HOME_DIRECTORY} ${CMAKE_HOME_D
 
 set(battleAI_SRCS
 		StdInc.cpp
+		${CMAKE_BINARY_DIR}/Version.cpp
```
Comment (ArseniyShestakov) [inline AI/BattleAI/CMakeLists.txt]: Oh yeah get rid of this line since it's no longer there in develop.

Comment (FeniksFire) [inline AI/BattleAI/CMakeLists.txt]: already removed ;)

**Comment by FeniksFire** [comment]
@alexvins so what now, when I corrected what you wanted?

File: AI/BattleAI/BattleAI.cpp [hunk 2]
```diff
@@ -1,88 +1,34 @@
-#include "StdInc.h"
```
Comment (alexvins) [inline AI/BattleAI/BattleAI.cpp]: This must be included in every .cpp file

File: AI/BattleAI/common.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+#include "../../CCallback.h"
```
Comment (alexvins) [inline AI/BattleAI/common.h]: No need to include whole file, forward declarations will be enough

File: AI/BattleAI/StackWithBonuses.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+#include "../../lib/HeroBonus.h"
+#include "../../lib/BattleState.h"
```
Comment (alexvins) [inline AI/BattleAI/StackWithBonuses.h]: should be CStack forward declaration.

File: AI/BattleAI/StackWithBonuses.cpp [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#include "StackWithBonuses.h"
+
+const TBonusListPtr StackWithBonuses::getAllBonuses(const CSelector &selector, const CSelector &limit,
+													const CBonusSystemNode *root /*= nullptr*/, const std::string &cachingStr /*= ""*/) const
```
Comment (alexvins) [inline AI/BattleAI/StackWithBonuses.cpp]: to much space here

File: AI/BattleAI/AttackPossibility.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ */
+
+#pragma once
+
+#include "common.h"
+#include "../../lib/BattleState.h"
+#include "../../lib/CCreatureHandler.h"
+#include "../../lib/spells/CSpellHandler.h"
+#include "../../lib/VCMI_Lib.h"
```
Comment (alexvins) [inline AI/BattleAI/AttackPossibility.h]: no includes here is needed actually. Use forward declarations.

File: AI/BattleAI/AttackPossibility.cpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
+#include "AttackPossibility.h"
```
Comment (alexvins) [inline AI/BattleAI/AttackPossibility.cpp]: Missing "banner". (In other .cpp also)

File: AI/BattleAI/BattleAI.cpp [hunk 8]
```diff
@@ -179,94 +125,9 @@ BattleAction CBattleAI::activeStack( const CStack * stack )
...
-
-void CBattleAI::battleStacksEffectsSet(const SetStackEffect & sse)
-{
-	print("battleStacksEffectsSet called");
-}
-
-void CBattleAI::battleStart(const CCreatureSet *army1, const CCreatureSet *army2, int3 tile, const CGHeroInstance *hero1, const CGHeroInstance *hero2, bool Side)
-{
-	print("battleStart called");
-	side = Side;
```
Comment (alexvins) [inline AI/BattleAI/BattleAI.cpp]: this line may be useful.

File: AI/BattleAI/BattleAI.cpp [hunk 9]
```diff
@@ -107,12 +53,12 @@ BattleAction CBattleAI::activeStack( const CStack * stack )
...
-	cbc = cb; //TODO: make solid sure that AIs always use their callbacks (need to take care of event handlers too)
+	setCbc(cb); //TODO: make solid sure that AIs always use their callbacks (need to take care of event handlers too)
+
 	try
 	{
-		print("activeStack called for " + stack->nodeName());
-		if(stack->type->idNumber == CreatureID::CATAPULT)
-			return useCatapult(stack);
+		//		if(stack->type->idNumber == CreatureID::CATAPULT)
+		//			return useCatapult(stack);
```
Comment (alexvins) [inline AI/BattleAI/BattleAI.cpp]: It is not safe to just comment out this lines.

File: AI/BattleAI/AttackPossibility.h (hunk 6 above)
Comment (FeniksFire) [inline AI/BattleAI/AttackPossibility.h]: I fix all your comments in last commit except this one, in this file: 
`#include "../../lib/BattleState.h"`
I hope that you will understand why ;).

File: AI/BattleAI/BattleAI.cpp [hunk 10]
```diff
@@ -1,8 +1,16 @@
...
+ * BattleAI.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "BattleAI.h"
+#include "StdInc.h"
```
Comment (alexvins) [inline AI/BattleAI/BattleAI.cpp]: StdInc must be first include

File: AI/BattleAI/EnemyInfo.h [hunk 11]
```diff
@@ -7,10 +7,11 @@
...
  *
  */
-
 #pragma once
-
-#include "ThreatMap.h"
+#include "common.h"
+#include "CCallback.h"
+#include "../../lib/BattleState.h"
+class CStack;
```
Comment (alexvins) [inline AI/BattleAI/EnemyInfo.h]: Do not include BattleState.h and decalre CStack at the same time. Prefere to use forward declaration. and use inlude only if it is absolutely necessary. 
Actually you need _only_ include BattleHex.h here.

File: AI/BattleAI/BattleAI.cpp [hunk 12]
```diff
@@ -298,8 +166,8 @@ BattleAction CBattleAI::goTowards(const CStack * stack, BattleHex destination)
 		{
 			auto nearestNeighbourToHex = vstd::minElementByFun(destNeighbours, [&](BattleHex a)
 			{
-				return BattleHex::getDistance(a, hex);
-			});
+						     return BattleHex::getDistance(a, hex);
```
Comment (ArseniyShestakov) [inline AI/BattleAI/BattleAI.cpp]: More spaces there.

File: AI/BattleAI/CMakeLists.txt [hunk 13]
```diff
@@ -6,7 +6,13 @@ include_directories(${Boost_INCLUDE_DIRS} ${CMAKE_HOME_DIRECTORY} ${CMAKE_HOME_D
 set(battleAI_SRCS
 		StdInc.cpp
 		BattleAI.cpp
+                StackWithBonuses.cpp
```
Comment (ArseniyShestakov) [inline AI/BattleAI/CMakeLists.txt]: I don't care that much about cmake files, but those are spaces too BTW.

File: AI/BattleAI/CMakeLists.txt [hunk 14]
```diff
@@ -10,7 +10,7 @@ set(battleAI_SRCS
                 EnemyInfo.cpp
                 AttackPossibility.cpp
                 PotentialTargets.cpp
-		main.cpp
+                main.cpp
```
Comment (FeniksFire) [inline AI/BattleAI/CMakeLists.txt]: Wtf, I don't know why but my Qt Creator show me tabs. I try commit again.

File: AI/BattleAI/BattleAI.cpp (hunk 8 above)
Comment (ArseniyShestakov) [inline AI/BattleAI/BattleAI.cpp]: Yeah it's looks like this one is useful. With new code CGameHandler::makeCustomAction complain about missing side in some cases.

**Comment by FeniksFire** [comment]
done, anything else?

**Comment by dydzio0614** [comment]
Rebase and glue commits into one? :P

**Comment by DjWarmonger** [comment]
Does this commit change any functionality beside refactoring? I see no point to merge it until major version release tomorrow.

**Comment by FeniksFire** [comment]
Ehh, this pull nothing change, really. Only divide BattleAI into smaller files for the purpose to do later tdd. I don't want change anything in this code, it's simply divided and commented part which vcmi doesn't use.

**Comment by ArseniyShestakov** [comment]
@DjWarmonger so you want major release soon? Can you please test serializer branch then?

**Pull request merged**